src/cluster_legacy.c

```diff
@@ -2990,7 +2991,51 @@ int clusterProcessPacket(clusterLink *link) {
 
                 if (clusterNodeIsMaster(sender)) {
                     /* Master turned into a slave! Reconfigure the node. */
-                    clusterDelNodeSlots(sender);
+                    if (master && !memcmp(master->shard_id, sender->shard_id, CLUSTER_NAMELEN)) {
+                        /* 'sender' was a primary and was in the same shard as `master`, its new primary */
+                        if (sender->configEpoch > senderConfigEpoch) {
+                            serverLog(LL_WARNING,
+                                    "Ignore stale gossip message from %.40s (%s) in shard %.40s;"
+                                    " gossip config epoch: %llu, current config epoch: %llu", 
+                                    sender->name,
+                                    sender->human_nodename,
+                                    sender->shard_id,
+                                    (unsigned long long)senderConfigEpoch,
+                                    (unsigned long long)sender->configEpoch);
+                        } else {
+                            /* A failover occurred in the shard where `sender` belongs to and `sender` is no longer
+                             * a primary. Update slot assignment to `master`, which is the new primary in the shard */
+                            int slots = clusterMoveNodeSlots(sender, master);
+                            serverLog(LL_NOTICE, "A failover occurred in shard %.40s; node %.40s (%s)"
+                                    " lost %d slot(s) to node %.40s (%s)",
+                                    sender->shard_id,
+                                    sender->name,
+                                    sender->human_nodename,
+                                    slots,
+                                    master->name,
+                                    master->human_nodename);
+                        }
+                    } else {
+                        int slots = clusterDelNodeSlots(sender);
+                        if (master != NULL) {
+                            /* `sender` was moved to another shard and has become a replica, remove its slot assignment */
```

Comment:
```suggestion
                        /* `sender` was moved to another shard and has become a replica, remove its slot assignment */

                        int slots = clusterDelNodeSlots(sender);
                        if (master != NULL) {
```
Seems like this comment should be higher? This is technically true in both cases, the other case we just also don't know the new master.

Reply:
good point.